Sources/Sentry/SentryNetworkTracker.m

```diff
@@ -0,0 +1,152 @@
+SentryNetworkTracker ()
+
+@property (nonatomic, strong) SentryPerformanceTracker *tracker;
+
+@end
+
+@implementation SentryNetworkTracker
+
++ (SentryNetworkTracker *)sharedInstance
+{
+    static SentryNetworkTracker *instance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{ instance = [[self alloc] init]; });
+    return instance;
+}
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        self.tracker = SentryPerformanceTracker.shared;
+    }
+    return self;
+}
+
+- (void)urlSessionTaskResume:(NSURLSessionTask *)sessionTask
+{
+    NSURL *url = [[sessionTask currentRequest] URL];
+
+    if (url == nil || ![self isTaskSupported:sessionTask])
+        return;
+
+    NSURL *apiUrl = [NSURL URLWithString:SentrySDK.options.dsn];
+    if ([url.host isEqualToString:apiUrl.host] && [url.path containsString:apiUrl.path])
+        return;
+
+    NSString *statePath = NSStringFromSelector(@selector(state));
+    [sessionTask addObserver:self
+                  forKeyPath:statePath
+                     options:NSKeyValueObservingOptionNew
+                     context:nil];
+
+    SentrySpanId *spanId =
+        [self.tracker startSpanWithName:[NSString stringWithFormat:@"%@ %@",
+                                                  sessionTask.currentRequest.HTTPMethod, url]
+                              operation:SENTRY_NETWORK_REQUEST_OPERATION];
+
+    objc_setAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID, spanId,
+        OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey, id> *)change
+                       context:(void *)context
+{
+    if ([keyPath isEqualToString:NSStringFromSelector(@selector(state))]) {
+        NSURLSessionTask *sessionTask = object;
+        if (sessionTask.state != NSURLSessionTaskStateRunning) {
+            SentrySpanId *spanId
+                = objc_getAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID);
+
+            if (spanId != nil) {
+                id<SentrySpan> span = [self.tracker getSpan:spanId];
+                NSInteger responseStatusCode = [self urlResponseStatusCode:sessionTask.response];
+
+                [span setDataValue:[NSNumber numberWithInteger:responseStatusCode]
+                            forKey:@"http.status_code"];
+
+                [self.tracker finishSpan:spanId withStatus:[self statusForSessionTask:sessionTask]];
+                [sessionTask removeObserver:self forKeyPath:NSStringFromSelector(@selector(state))];
+            }
+        }
+    }
+}
+
+- (NSInteger)urlResponseStatusCode:(NSURLResponse *)response
+{
+    if (response != nil && [response isKindOfClass:[NSHTTPURLResponse class]]) {
+        return ((NSHTTPURLResponse *)response).statusCode;
+    }
+    return -1;
+}
+
+- (SentrySpanStatus)statusForSessionTask:(NSURLSessionTask *)task
```

Comment:
`h`: The code doesn't check if there is an `error` set on the `task`. When you are for example offline this code would now end the span successfully.